openmicroscopy/openmicroscopy
# Shapedata bugfix

# What this PR does

Allows to remove the C, Z and T of a Shape. Previously when a negative value was passed to `ShapeData.setC()` (etc.) the value was set to `0`.  With this PR passing a negative value will set the underlying Ice Shape object's C, Z, T to `null`.  And for the other way round, if the Shape object's C, Z, T is `null`, `ShapeData.getC()` (etc.) will return `-1`, indicating that the value has not been set, so that getters and setters now also behave in a consistent way.

Had to modify the `ROIData` pojo, too, e.g. remove methods like `firstPlane()` (which isn't used anywhere), because these wouldn't work properly with `null` CZT ROIs (which are supposed to be present on all planes).

Also updated the integration test accordingly.

Another bug was noticed: The various setters (`setX()`, `setY()`, etc.) on the ShapeData subclasses (`RectangleData`, etc.) didn't set the `dirty` flag, that's why the ROIFacility didn't save them back to the server when `saveROIs` was called. Fixed the setters and added an unit test for all setters of all ShapeData classes.

# Testing this PR

- Check ROIFacility integration test.
- Check ShapeData unit test.
- Check that ROI tool in Insight still works as before.

# Related reading

http://lists.openmicroscopy.org.uk/pipermail/ome-devel/2018-May/004199.html

/cc @awalter17
